### PR TITLE
Fix shmem allocation size.

### PR DESCRIPTION
**Pull request by rjuju**
MaxBackends is still 0 when _PG_init() is called, which means that we
don't request enough memory in RequestAddinShmemSpace(), while the rest of
the code sees (and allocate) a correct value.

It's technically usually not a problem as postgres adds an extra 100kB of
memory for small unaccounted memory usage, but it's better to avoid relying on
it too much.

Note that the value is still not guaranteed to be exact as other modules
_PG_init() could later change the underlying GUCs, but there is not available
API to handle that case accurately.

### CODE REVIEW

**Comment by maksm90** [comment]
@rjuju thanks for so swift fix of https://github.com/postgrespro/pg_wait_sampling/issues/41 . Review it shortly

File: pg_wait_sampling.c [hunk 1]
```diff
@@ -75,7 +79,12 @@ get_max_procs_count(void)
 	int count = 0;
 
 	/* MyProcs, including autovacuum workers and launcher */
-	count += MaxBackends;
+	count += MaxConnections + autovacuum_max_workers + 1
+			+ max_worker_processes
+#if PG_VERSION_NUM >= 120000
+			+ max_wal_senders
+#endif
+			+ 1;
```
Comment (maksm90) [inline pg_wait_sampling.c:87]: To which kind of process does belong the last one? `MaxBackends` in [PG15](https://github.com/postgres/postgres/blob/002c9dd97a0c874fd1693a570383e2dd38cd40d5/src/backend/utils/init/postinit.c#L555-L557) and before is calculated without this.

**Comment by maksm90** [comment]
Some notes about the relevance of this kind of estimation of max processes in `pg_wait_sampling` module.

- This value is used to allocate shared memory just for storing queryId (in `proc_queryids` array) as in `pg_stat_kcache` module. But instead of `pg_stat_kcache` here we assign one slot for queryId to one slot of PGPROC therefore we have to use formula for the size of `ProcGlobal->allProcs` array. And the last sum item `max_prepared_xacts` might be removed as we access to `proc_queryids` array from regular backends and iterate over it up to `ProcGlobal->allProcCount` index value that is without this item.
Although it's enough to allocate `MaxBackends`-sized array and access to this array using `MyBackendId-1` index value similar as it's [done](https://github.com/powa-team/pg_stat_kcache/blob/master/pg_stat_kcache.c#L417) in `pg_stat_kcache` module.
-  Starting from PG14 we might use `queryId` value from shared `BackendStatusArray` and don't allocate own shared memory for `queryId` values, in general. @rjuju what do you think what if we rewrote the storing and extracting of `queryId` in such way in future?

File: pg_wait_sampling.c (hunk 1 above)
Comment (maksm90) [reply pg_wait_sampling.c:87]: We don't access to the array for `queryId` values using `MyBackendId` index value. Canonical way to define own index is `MyProc - ProcGlobal->allProcs`

Comment (rjuju) [reply pg_wait_sampling.c:87]: Yes, sorry the extra +1 comes from pg_stat_kcache code where we use a 1-based array access.  I wanted to remove it and forgot as I was focusing on the v12 change.  I will fix it!

**Comment by rjuju** [comment]
>  And the last sum item max_prepared_xacts might be removed as we access to proc_queryids array from regular backends and iterate over it up to ProcGlobal->allProcCount index value that is without this item.

Agreed, and the code should have a comment saying that the value has to be synced with ProcGlobal->allProcCount initialization in InitProcGlobal(), so that it's clear to anyone looking at the code why this value is used.

> Starting from PG14 we might use queryId value from shared BackendStatusArray and don't allocate own shared memory for queryId values, in general. @rjuju what do you think what if we rewrote the storing and extracting of queryId in such way in future?

Unfortunately it's not possible :(  The community insisted to report only the top-level queryid there, for consistency with e.g. pg_stat_activity.query, even if I initially suggested to report the current query.  So we, extension owners, still have to allocate and handle our own queryid array.  Unfortunately there isn't even a simple way to do it only once and not once per extension.

**Comment by maksm90** [comment]
> the code should have a comment saying that the value has to be synced with ProcGlobal->allProcCount initialization in InitProcGlobal(), so that it's clear to anyone looking at the code why this value is used.

Yeah, it's a good idea to document the formula in that way.

> Unfortunately it's not possible :( The community insisted to report only the top-level queryid there, for consistency with e.g. pg_stat_activity.query, even if I initially suggested to report the current query. So we, extension owners, still have to allocate and handle our own queryid array.

It's a pity.

>  Unfortunately there isn't even a simple way to do it only once and not once per extension.

Yes, in essence, multiple extensions have to maintain its own storage for `queryId`s of currently executing queries for sharing these values between processes. And clearly, this is redundantly.

**Comment by rjuju** [comment]
I just force-pushed the modifications we discussed.  I tried to describe the problem with all consequences, so that if anyone tries to read or use the code for something else they will have all the information needed to know if that approach will work for them too.

**Review by maksm90 — APPROVED** [review]
> I just force-pushed the modifications we discussed. I tried to describe the problem with all consequences, so that if anyone tries to read or use the code for something else they will have all the information needed to know if that approach will work for them too.

Excellent! Thanks a lot. Approved.

**Comment by rjuju** [comment]
Thanks!

**Pull request merged**